### PR TITLE
[WIP] Glossary Page #1127

### PR DESCRIPTION
**Pull request by wackerow**
## Description
The beginnings of a **glossary** page. Kept simple in order to get something on the screen.

Utilized the Mastering Ethereum book glossary as a starting point as it is open-source, shared under CC-BY-SA (must credit them, and must share again) and also, it has already been curated and slightly tailored for Ethereum, rather than more Bitcoin centered. This can of course be expanded upon. 

Given the nature of these words, it felt like it belonged with developer tools (obviously can be changed as well). This list was parsed and saved into `src/data/developer-glossary.yaml`. Imported into new page `src/pages-conditional/developers/glossary.js` and mapped out for display.

Didn't have a sketch to go off, so obviously this will require reworking, but I did want to get _something_ drafted for this issue, as the comments seemed to quickly veer off into fancier glossary tools and SEO... The original feature was requested to have a simple place on ethereum.org to reference for a simple definition of some esoteric terms, and my goal here is to at least get something up that can then be discussed, tweaked, and expanded upon.

_**This currently still needs to be linked to from other pages/components_, but for now if you fire up `/en/developers/glossary/` on a local build of this branch it'll load. 

### Top of page
image
### Bottom of page
image

I wanted to seek some input before continuing too much further. Let me know what you think!

## Related Issue #1127

### CODE REVIEW

**Comment by ryancreatescopy** [comment]
Hey @wackerow this looks great. A few comments...

**From an IA perspective**
We've had some discussions internally about where best to place this content. I think our current thinking is that a glossary has wider usefulness than just for developers, so perhaps it should live outside the developers section. Would be keen for an outside perspective too though. I understand your point about the glossary in its current state (based on Mastering Ethereum)  but we'd eventually want to become less reliant on these definitions and perhaps make them more accessible to non-technical audiences.

**From a UX/UI perspective**
I'm hesitant to re-use the UI from the tutorials page. Simply because this looks to me as if each glossary entry is clickable – although I assume this won't be the case for every entry. We may also want to include a glossary search or a "filter by letter" to make it easier for users to find what they're looking for. MDN separates entries out by letter (https://developer.mozilla.org/en-US/docs/Glossary) but I think we could do one better and offer filtering. I understand this is a pretty hefty job so totally understand if you're not up for helping figure this out – you've left us a great foundation either way!

**From a content perspective** 
Related to the IA point above, ideally we'd want the glossary to be as accessible as possible. I've been wondering about whether entries could have an ELI5 (Explain like I'm 5) definition as default and then more technical terms could have a developer-centric explanation for users that want a more in-depth technical overview. On this note, we should think about how to link users to more info on a topic. For example, `account` – we can provide an ELI5 definition, a developer definition and then a link to the docs page on accounts. The content design of the "entry" would need to work for these content requirements. This might be something I can help with in terms of sketching out a new design soon – will bring this up in our sprint planning. 

My current thinking is that an entry should have the potential for:
- title
- ELI5 definition
- developer definition
- Links (whether to internal or external docs that can support learning)

@samajammin any more to add on this from a technical perspective? Are there any requirements if we do want to eventually implement glossary tooltips?

Welcome feedback on any of these thoughts and thanks again for jumping into these issues. Feel free to explore any of these ideas further or leave to us!

**Comment by wackerow** [comment]
@ryancreatescopy 
- "_From an IA perspective_": Totally agree on expanding this to non-technical users! I didn't want to be too obtrusive with the location of this new page off the bat, but if you have a better home for it in mind let me know! (I'll mull this over too as I work on it more)
- "_From a UX/UI perspective_": Yeah, it was mirrored after Tutorials page as a starting point (but completely with the intent of getting your feedback and overhauling it lol), and I agree about the highlighting as this was something left in place just as a "hey, maybe we could use this somehow for consistency" while this is a WIP. Also agree with a filter input field 👍 (vs a filter-by-letter selection 👎)... preferably one that freezes to the top of the screen as you scroll down. 
- "_From a content perspective_": I do like the idea of expanding to have ELI5 definition, of course this will require more content creation.
- For all of these, I would also want to implement an easy and encouraging way for the 'experts' to suggest new terms / edit existing terms, as this will help distribute the content creation portion.
-- Thoughts on best way to do this? Easiest (for us) would probably just be a link to here, but easiest for the user feels more like a simple form they can submit that then gets reviewed on this end. 
- "you've left us a great foundation"... thank you! That was my main goal with this, to just set a foundation... something to get the ideas flowing and something we can start building on. 

I can keep working on this and keep you updated here, but obviously feel free to dive in yourself or just let me know what you want

**Comment by samajammin** [comment]
Thanks for kicking off this work @wackerow! Super exciting to see.

I agree with you points @ryancreatescopy. A few thoughts:

Re: information architecture, I also think it should live outside the Developers section, e.g. ethereum.org/en/glossary

The MDN example is interesting. Do we think we should have dedicated pages for each term, like MDN? That would be better for SEO but for glossary terms that don't require much content/explanation, requiring an extra click from a glossary index page might actually hurt UX. Including a ELI5 as well as a more in depth definition & links might justify standalone pages.

**For a 1st version of this...**
I'd lean towards creating this as a single markdown file. This would give us a table of contents out of the box, so we could have anchor links by term / for each letter of the alphabet that users could easily skip to. While we couldn't easily build the filter functionality with this simple approach, people could still `command + F` to search the page & these terms would also show up in our site search results (creating each term as a header would trigger Algolia to index it). As the content grows, we could choose to break up terms into individual pages later.

**For future versions of this...**
I do still love the idea of glossary tooltips for terms (as discussed in #1127). Implementing that would certainly require a more complex structure to store each term & its definition but I don't think we should build for that yet. We can always refactor. I'd rather get a working glossary up ASAP since we know it will be immediately valuable.

Let me know what you folks think!

**Comment by samajammin** [comment]
@wackerow minor housekeeping note - Github provides the ability to open "draft" PRs, which is useful for PRs we know are not yet in a state to merge:
- https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request#creating-the-pull-request
- https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft

I've converted this one already 😄

**Comment by wackerow** [comment]
@samajammin very helpful, thanks for letting me know! I'll take all these into account and see what I can do

**Comment by wackerow** [comment]
Before I forget... was just browsing the MDN glossary pages, and one thing I like on their page is the "Learn a new term" card to emphasize a particular term (could be used elsewhere on site). Just a note for the future... thought this may be a nice feature to build once glossary is nailed down.

**Comment by wackerow** [comment]
#### As of 1d8b124
image
image
image

**Review by samajammin — DISMISSED** [review]
This looks terrific! Thanks @wackerow ❤️ 

@ryancreatescopy I defer to you on copy/design but I think this is good to merge for now. We can continue to iterate on improvements in future PRs.

**Comment by ryancreatescopy** [comment]
When I preview this deploy the content doesn't wrap on mobile. @samajammin how does this look for you on mobile?

**Comment by wackerow** [comment]
@samajammin Just noticed a layout bug on mobile:
image

I'll look at this now if you wanna hold off

**Comment by wackerow** [comment]
@samajammin @ryancreatescopy 🤔 Do either of you know how to fix this mobile styling issue? I'm confused because using the markdown approach and the static template I didn't have to do any styling to begin with... everything else fell into place, except this. Can't figure out why it's treating this one content file differently.

**Comment by ryancreatescopy** [comment]
Hmmm, my hunch may be that there's a bug when titles get too long for the table of contents... there are some very long titles in the glossary entries. But that's only a guess without taking a proper look

**Comment by wackerow** [comment]
@ryancreatescopy You had the right idea, it was the displayed link names that couldn't be broken up by spaces. I was definitely not thinking in that direction so thanks a lot! Changed it to be more generic and compact with "Read more at domain.com" format, fixes it beautifully.
I left the long names (i.e. "Ethereum Virtual Machine (EVM)") intentionally for search indexing purposes since that wasn't causing any layout problem. 
image

@samajammin Should be set for one last review!

**Review by ryancreatescopy — CHANGES_REQUESTED** [review]
This looks great, so close but some things to do before merging IMO:

1) Link to relevant pages for more info if a page exists, e.g. account, EIPs, transactions
2) Remove names like Satoshi and Vitalik and references to bitcoin terms, e.g. BIPs (not sure it's needed for our context)
3) Where an entry refers to a term that's also an entry we should provide an anchor link, e.g. entry for blockchain refers to "proof-of-work"

@wackerow you've put in so much work so totally understand if you want us to get this over the line, but if you're happy to, it's good to merge once these changes are made.

**Comment by wackerow** [comment]
@ryancreatescopy Oh I can definitely make these adjustments... the linking I planned to do, but yesterday we seemed to hit a "first version", so I was going to wait, but I can do it now.

So my only thought on the bitcoin terms... I totally agree for terms that have nothing to do with Ethereum, but some I would argue do, such as BIPs... (ie. BIP32 and BIP39 for HD wallets and mnemonics used in Ethereum). I considered removing these right off the bat, but held off since some of these terms are relevant to understanding Ethereum as well. 
 
Lemme know your thoughts, in the meantime I'll get this linked up.

**Comment by samajammin** [comment]
Thanks @wackerow! I think it's safe to remove Bitcoin-specific terms. Agree with you that aspects of Ethereum have been influenced by Bitcoin (e.g. BIP32 & BIP39) but I don't think these Bitcoin glossary terms make that any more clear - someone would still need to do some digging to find that info.

**Comment by samajammin** [comment]
Another glossary source (that can wait til a future PR): https://ethdocs.org/en/latest/glossary.html

License info: https://github.com/ethereum/homestead-guide/blob/master/LICENSE-CC-BY-SA-4-POINT-0

**Comment by wackerow** [comment]
@samajammin Sounds good!... Just pushed my last save point if you want to peek at the changes (through about a-e so far) which includes what I did to BIP... If I don't hear further I will just remove that when I review again.

**Review by samajammin — APPROVED** [review]
This is amazing! Thanks again 🎉

**Comment by samajammin** [comment]
@ryancreatescopy looks like you also need to approve in order to merge, since you requested changes ealier.

**Comment by ryancreatescopy** [comment]
Are swarm and whisper still relevant for Ethereum @samajammin @wackerow — I forget whether they're still an Ethereum thing... other than that small Q it looks great @wackerow !!!

**Comment by wackerow** [comment]
@ryancreatescopy Not sure about whisper, but I thought swarm still was...
@samajammin If you'd like 'swarm' and/or 'whisper' to be removed just let me know, otherwise ca886a8 should be ready to pull (added just a few last links to other parts of the website).

**Comment by wackerow** [comment]
sorry! made a formatting typo with one markdown link that I couldn't leave in there (fixed in e336d88)
phew... now _that_ should do it 🤞🏻

**Comment by samajammin** [comment]
Good point @ryancreatescopy ... perhaps we should make a policy to not include specific projects in the glossary as well. Truffle for instance. It feels weird to add them but not other projects. That may create a slippery slope of promotion.

**Comment by wackerow** [comment]
This looks good from my end after all the tweaks (and a little more time to look at it)... Gave it a break yesterday after embarrassingly having to commit two typo fixed lol, sorry... I'm leaving any other potential improvements until after pulled, unless you have any other immediate fixes in mind. @samajammin and @ryancreatescopy I thank you both for including me in this!

File: src/content/glossary/index.md
Comment (samajammin) [inline src/content/glossary/index.md:99]:
```suggestion
### difficulty bomb {#difficulty-bomb}
```

File: src/intl/en.json
Comment (samajammin) [inline src/intl/en.json:82]:
```suggestion
  "page-glossary": "Ethereum Glossary",
```

**Review by samajammin — APPROVED** [review]

**Review by ryancreatescopy — APPROVED** [review]

**Pull request merged**
